gossip: use atomic value for ready channel

kvoli opened the pull request:
The ready channel is used to signal waiting gossip requests. Previously, the gossip mutex was held when signaling the channel which led to contention.

Use an atomic instead of holding the gossip mutex when signaling the ready channel.

Epic: none
Release note: None

kvoli committed:
gossip: use atomic value for ready channel

The ready channel is used to signal waiting gossip requests. Previously,
the gossip mutex was held when signaling the channel which led to
contention.

Use an atomic instead of holding the gossip mutex when signaling the
ready channel.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126571)
<!-- Reviewable:end -->

nvanbenschoten commented on pkg/gossip/server.go:
```diff
@@ -370,11 +373,10 @@ func (s *server) start(addr net.Addr) {
-		ready := make(chan struct{})
-		close(s.mu.ready)
-		s.mu.ready = ready
+		ready := s.ready.Load().(chan struct{})
```
I was wrong about the order of closing and storing. We actually want the opposite order, so that the load of the channel after the previous one is signaled is guaranteed to read the new channel. So something like:
```go
close(s.ready.Swap(make(chan struct{})).(chan struct{}))
```

kvoli committed:
gossip: don't unlock mutex when sending reply

The gossip function was needlessly locking the gossip mutex for sending
a message, only to quickly re-acquire it. Avoid unlocking as we expect
the send duration to be minimal and contention reduction from holding
the lock throughout to be beneficial.

Epic: none
Release note: None

nvb reviewed (changes requested):
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126571)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/gossip/server.go` line 153 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126571#-O0ordii98zF_WFmqhx-:-O0ordii98zF_WFmqhx0:bodtawg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1842c08ade59c3f26cbfdef243e5972ae4834beb/pkg/gossip/server.go#L153)):*
> ```Go
> 
> 	for init := true; ; init = false {
> 		// Store the old ready so that if it gets replaced with a new one
> ```

nit: it looks wrong to see "Store" next to an atomic load. Consider replacing with "Remember".
___
*[`pkg/gossip/server.go` line 154 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126571#-O0os-gYBMBqkj0tg6w5:-O0os-gYBMBqkj0tg6w6:b-8l6gfr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1842c08ade59c3f26cbfdef243e5972ae4834beb/pkg/gossip/server.go#L154)):*
> once the lock is released

This part of the comment needs an update, now that we're loading outside of the lock.
___
*[`pkg/gossip/server.go` line 188 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126571#-O0osOEUAWF-X1HrMuyI:-O0osOEUAWF-X1HrMuyJ:b-2g1tmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/083ee356d9210474d2edee9f46e6f320f53b19c3/pkg/gossip/server.go#L188)):*
> ```Go
> 
> 			if err := send(reply); err != nil {
> 				s.mu.Unlock()
> ```

We don't want to hold the lock across the call to `send`, which can block on gRPC.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126571)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/gossip/server.go` line 376 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126571#gh-1662919885:-O0p0y9h0GNcPRrJRzMK:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b28ea5f99642434570e4742a452caa0f679f4/pkg/gossip/server.go#L376)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I was wrong about the order of closing and storing. We actually want the opposite order, so that the load of the channel after the previous one is signaled is guaranteed to read the new channel. So something like:
```go
close(s.ready.Swap(make(chan struct{})).(chan struct{}))
```
</blockquote></details>

Updated.
___
*[`pkg/gossip/server.go` line 153 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126571#-O0ordii98zF_WFmqhx-:-O0ougmk8zmsBqJ_hjHh:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/1842c08ade59c3f26cbfdef243e5972ae4834beb/pkg/gossip/server.go#L153)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: it looks wrong to see "Store" next to an atomic load. Consider replacing with "Remember".
</blockquote></details>

Updated.
___
*[`pkg/gossip/server.go` line 154 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126571#-O0os-gYBMBqkj0tg6w5:-O0ouhTpASoM-X9xLodd:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/1842c08ade59c3f26cbfdef243e5972ae4834beb/pkg/gossip/server.go#L154)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> once the lock is released

This part of the comment needs an update, now that we're loading outside of the lock.
</blockquote></details>

Updated.
___
*[`pkg/gossip/server.go` line 188 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126571#-O0osOEUAWF-X1HrMuyI:-O0p0z3JBqS3FdNDpVhG:b-1ut1fq) ([raw file](https://github.com/cockroachdb/cockroach/blob/083ee356d9210474d2edee9f46e6f320f53b19c3/pkg/gossip/server.go#L188)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We don't want to hold the lock across the call to `send`, which can block on gRPC.
</blockquote></details>

I misunderstood the earlier comment - updated to unblock on the else path here and also before send.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126571#-:-O0pEf8A5SGX4bhsQc6z:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @kvoli)



<!-- Sent from Reviewable.io -->

kvoli commented:
TYFTR!

kvoli commented:
bors r=nvanbenschoten
